R/JHU.R

```diff
@@ -0,0 +1,117 @@
+      "daily_recovered" = "https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_recovered_global.csv"
+    ),
+    # nolint end
+    #' @field source_data_cols existing columns within the raw data
+    source_data_cols = c("confirmed", "deaths", "recovered"),
+
+    #' @description Set up a table of region codes for clean data
+    #' @importFrom tibble tibble
+    set_region_codes = function() {
+      self$codes_lookup$`1` <- JHU_codes
+    },
+
+    #' @description JHU specific data cleaning
+    #' @importFrom dplyr last_col bind_rows mutate rename select everything
+    #' @importFrom tidyr pivot_longer pivot_wider replace_na
+    #' @importFrom lubridate mdy
+    #' @importFrom rlang .data
+    clean_common = function() {
+      self$data$clean <- lapply(self$data$raw,
+        pivot_longer,
+        cols = 5:last_col(),
+        names_to = "Date",
+        values_to = "value"
+      )
+      self$data$clean <- self$data$clean %>%
+        bind_rows(.id = "variable") %>%
+        pivot_wider(names_from = variable, values_from = value) %>%
+        select(
+          .data$Date,
+          .data$`Province/State`,
+          .data$`Country/Region`,
+          .data$daily_confirmed,
+          .data$daily_deaths,
+          .data$daily_recovered,
+          everything()
+        ) %>%
+        mutate(
+          Date = mdy(.data$Date),
+          daily_confirmed = as.numeric(.data$daily_confirmed),
+          daily_deaths = as.numeric(.data$daily_deaths),
+          daily_recovered = as.numeric(.data$daily_recovered),
+          level_2_region_code = NA_real_
+        ) %>%
+        rename(
+          date = .data$Date,
+          level_2_region = .data$`Province/State`,
+          level_1_region = .data$`Country/Region`,
+          cases_total = .data$daily_confirmed,
+          deaths_total = .data$daily_deaths,
+          recovered_total = .data$daily_recovered
+        ) %>%
+        replace_na(
+          list(
+            level_1_region = "Unknown",
+            level_2_region = "Unknown"
+          )
+        ) %>%
+        full_join(
+          self$codes_lookup$`1`,
+          by = c("level_1_region" = "Country_Region")
+        )
+    },
+
+    #' @description JHU specific country level data cleaning
+    #' @importFrom dplyr select summarise group_by across everything
+    #' @importFrom rlang .data
+    clean_level_1 = function() {
+      self$data$clean <- self$data$clean %>%
+        select(
+          .data$date, .data$level_1_region_code, .data$level_1_region,
+          .data$cases_total, .data$deaths_total, .data$recovered_total,
+          everything(), -.data$level_2_region_code
+        ) %>%
+        group_by(
+          .data$date, .data$level_1_region_code,
+          .data$level_1_region
+        ) %>%
+        summarise(across(where(is.double), sum))
```

Comment:
are there any the data? at the moment this will propagate them upwards which is fine but worth thinking about.

Reply:
I think there are, should I use `na.rm = TRUE` here?